src/test/java/org/opensearch/flowframework/transport/CreateWorkflowTransportActionTests.java

```diff
@@ -146,35 +153,59 @@ public void testFailedDryRunValidation() {
 
         createWorkflowTransportAction.doExecute(mock(Task.class), createNewWorkflow, listener);
         ArgumentCaptor<Exception> exceptionCaptor = ArgumentCaptor.forClass(Exception.class);
         verify(listener, times(1)).onFailure(exceptionCaptor.capture());
         assertEquals("No start node detected: all nodes have a predecessor.", exceptionCaptor.getValue().getMessage());
     }
 
+    /* public void testMaxWorkflow() {
+        ActionListener<WorkflowResponse> listener = mock(ActionListener.class);
+        ArgumentCaptor<ActionListener<Boolean>> actionListenerCaptor = ArgumentCaptor.forClass(ActionListener.class);
+        WorkflowRequest createNewWorkflow = new WorkflowRequest(null, template, null, 1);
+
+        doAnswer(invocation -> {
+            ActionListener<Boolean> responseListener = invocation.getArgument(1);
+            responseListener.onResponse(true);
+            return null;
+        }).when(mock(CreateWorkflowTransportAction.class)).onSearchGlobalContext(any(), any(), actionListenerCaptor.capture());
+
+        createWorkflowTransportAction.doExecute(mock(Task.class), createNewWorkflow, listener);
+        ArgumentCaptor<Exception> exceptionCaptor = ArgumentCaptor.forClass(Exception.class);
+        verify(listener, times(1)).onFailure(exceptionCaptor.capture());
+        assertEquals("Unable to fetch the workflows", exceptionCaptor.getValue().getMessage());
+    }
+
     public void testFailedToCreateNewWorkflow() {
         @SuppressWarnings("unchecked")
         ActionListener<WorkflowResponse> listener = mock(ActionListener.class);
-        WorkflowRequest createNewWorkflow = new WorkflowRequest(null, template);
+        ArgumentCaptor<ActionListener<Boolean>> actionListenerCaptor = ArgumentCaptor.forClass(ActionListener.class);
+        WorkflowRequest createNewWorkflow = new WorkflowRequest(null, template, null, 1);
 
         doAnswer(invocation -> {
             ActionListener<IndexResponse> responseListener = invocation.getArgument(1);
             responseListener.onFailure(new Exception("Failed to create global_context index"));
             return null;
         }).when(flowFrameworkIndicesHandler).putTemplateToGlobalContext(any(Template.class), any());
 
+        doAnswer(invocation -> {
+            ActionListener<SearchResponse> responseListener = invocation.getArgument(1);
+            responseListener.onResponse(new SearchResponse(null));
+            return null;
+        }).when(mock(CreateWorkflowTransportAction.class)).onSearchGlobalContext(any(), any(), actionListenerCaptor.capture());
+
         createWorkflowTransportAction.doExecute(mock(Task.class), createNewWorkflow, listener);
         ArgumentCaptor<Exception> exceptionCaptor = ArgumentCaptor.forClass(Exception.class);
         verify(listener, times(1)).onFailure(exceptionCaptor.capture());
         assertEquals("Failed to create global_context index", exceptionCaptor.getValue().getMessage());
-    }
+    }*/
 
```

Comment:
Not sure I understand why we need to wait until integration tests to test these methods. These methods accept an action listener, we can definitely stub them to invoke `onFailure` or `onResponse` and assert the exception messages/ response values

Reply:
The same reason why we commented the tests below before. Now, we have 2 actionListeners and stubbing them is making the test complex. It will be better to test these functionalities using integ test. That's how other plugins are doing today. Open to suggestions though.

Reply:
Took a stab at it and got a working test for the successful creation of a workflow : 
In order to stub the `createWorkflowTransportAction.checkMaxWorkflows()` we just need to `spy()` the object. like so : 
```java
this.createWorkflowTransportAction = spy(
            new CreateWorkflowTransportAction(
                mock(TransportService.class),
                mock(ActionFilters.class),
                workflowProcessSorter,
                flowFrameworkIndicesHandler,
                settings,
                client
            )
        );
```

And then we can stub all the nested calls
```java
 public void testCreateNewWorkflow() {

        @SuppressWarnings("unchecked")
        ActionListener<WorkflowResponse> listener = mock(ActionListener.class);
        WorkflowRequest workflowRequest = new WorkflowRequest(
            null,
            template,
            false,
            WORKFLOW_REQUEST_TIMEOUT.get(settings),
            MAX_WORKFLOWS.get(settings)
        );

        // Bypass checkMaxWorkflows and force onResponse
        doAnswer(invocation -> {
            ActionListener<Boolean> checkMaxWorkflowListener = invocation.getArgument(2);
            checkMaxWorkflowListener.onResponse(true);
            return null;
        }).when(createWorkflowTransportAction).checkMaxWorkflows(any(TimeValue.class), anyInt(), any());

        // Bypass putTemplateToGlobalContext and force onResponse
        doAnswer(invocation -> {
            ActionListener<IndexResponse> responseListener = invocation.getArgument(1);
            responseListener.onResponse(new IndexResponse(new ShardId(GLOBAL_CONTEXT_INDEX, "", 1), "1", 1L, 1L, 1L, true));
            return null;
        }).when(flowFrameworkIndicesHandler).putTemplateToGlobalContext(any(), any());

        // Bypass putInitialStateToWorkflowState and force on response
        doAnswer(invocation -> {
            ActionListener<IndexResponse> responseListener = invocation.getArgument(2);
            responseListener.onResponse(new IndexResponse(new ShardId(WORKFLOW_STATE_INDEX, "", 1), "1", 1L, 1L, 1L, true));
            return null;
        }).when(flowFrameworkIndicesHandler).putInitialStateToWorkflowState(any(), any(), any());

        ArgumentCaptor<WorkflowResponse> workflowResponseCaptor = ArgumentCaptor.forClass(WorkflowResponse.class);

        createWorkflowTransportAction.doExecute(mock(Task.class), workflowRequest, listener);

        verify(listener, times(1)).onResponse(workflowResponseCaptor.capture());

        assertEquals("1", workflowResponseCaptor.getValue().getWorkflowId());
    }
```

Reply:
This helped. Thanks @joshpalis. I have added all the tests and got the commented one working as well. `spy` was the blocker here.